lib/Frontend/nvqpp/ConvertExpr.cpp

```diff
@@ -1605,6 +1613,77 @@ bool QuakeBridgeVisitor::VisitCallExpr(clang::CallExpr *x) {
+    // Extract number of targets
+    std::size_t targetCount = 0;
+    std::string genFuncName;
+    for (auto name : customOperationNames) {
+      if (name.first.find(maybeUnitaryGenerator) != std::string::npos) {
+        targetCount =
+            std::stoi(name.first.substr(maybeUnitaryGenerator.length()));
+        genFuncName = getCudaqKernelName(name.second);
+        break;
+      }
+    }
+    if (targetCount > 0) {
+      // Extract number of parameters
+      std::size_t paramCount = [&]() {
+        std::size_t count = 0;
+        for (auto arg : args) {
+          auto argTy = arg.getType();
+          if (isa<FloatType>(argTy)) {
+            count++;
+          } else if (auto ptrTy = dyn_cast<cc::PointerType>(argTy)) {
+            if (isa<FloatType>(ptrTy.getElementType()))
+              count++;
+          }
+        }
+        return count;
+      }();
+      auto srefAttr = SymbolRefAttr::get(
+          StringAttr::get(builder.getContext(), genFuncName));
+      ValueRange operands(args);
+      assert(operands.size() >= 1 && "must be at least 1 operand");
+      if ((operands.size() == 1) &&
+          operands[0].getType().isa<quake::VeqType>()) {
+        auto target = operands[0];
+        if (!negations.empty())
+          reportNegateError();
+        Type i64Ty = builder.getI64Type();
+        auto size = builder.create<quake::VeqSizeOp>(
+            loc, builder.getIntegerType(64), target);
+        Value rank = builder.create<cudaq::cc::CastOp>(
+            loc, i64Ty, size, cudaq::cc::CastOpMode::Unsigned);
+        auto bodyBuilder = [&](OpBuilder &builder, Location loc, Region &,
+                               Block &block) {
+          Value ref = builder.create<quake::ExtractRefOp>(loc, target,
+                                                          block.getArgument(0));
+          builder.create<quake::CustomUnitarySymbolOp>(loc, srefAttr,
+                                                       ValueRange(), ref);
+        };
+        cudaq::opt::factory::createInvariantLoop(builder, loc, rank,
+                                                 bodyBuilder);
+      } else {
+        auto [targets, ctrls] =
+            maybeUnpackOperands(builder, loc, operands.drop_front(paramCount),
+                                isControl, targetCount);
+        for (auto v : targets)
+          if (std::find(negations.begin(), negations.end(), v) !=
+              negations.end())
+            reportNegateError();
+        auto negs =
+            negatedControlsAttribute(builder.getContext(), ctrls, negations);
+        SmallVector<Value> params;
+        for (auto p : operands.take_front(paramCount)) 
+          if (p.getType().isa<cudaq::cc::PointerType>()) 
+            params.push_back(builder.create<cudaq::cc::LoadOp>(loc, p));
+        builder.create<quake::CustomUnitarySymbolOp>(
+            loc, srefAttr, isAdjoint, params, ctrls, targets, negs);
+      }
+      return true;
+    }
+
```

Comment:
Where is the handling of complex expressions?

Reply:
Will be handled later.

Reply:
Does this mean in a separate PR? (This PR is already overly large, so I hope the answer is yes.)

Reply:
yes, I will add in follow-up PR.